### PR TITLE
[HEXAGON] AddrModeOpt support for HVX and optimize adds

### PR DESCRIPTION
**Pull request by quic-asaravan**
This patch does 3 things:
1. Add support for optimizing the address mode of HVX load/store instructions
2. Reduce the value of Add instruction immediates by replacing with the difference from other Addi instructions that share common base:

   For Example, If we have the below sequence of instructions: r1 = add(r2,# 1024) ... r3 = add(r2,# 1152) ... r4 = add(r2,# 1280)

   Where the register r2 has the same reaching definition, They get modified to the below sequence:

       r1 = add(r2,# 1024)
            ...
       r3 = add(r1,# 128)
            ...
       r4 = add(r1,# 256)
3. Fixes a bug pass where the addi instructions were modified based on a predicated register definition, leading to incorrect output.

Eg:
         INST-1: if (p0) r2 = add(r13,# 128)
         INST-2: r1 = add(r2,# 1024)
         INST-3: r3 = add(r2,# 1152)
         INST-4: r5 = add(r2,# 1280)

In the above case, since r2's definition is predicated, we do not want to modify the uses of r2 in INST-3/INST-4 with add(r1,#128/256)

4.Fixes a corner case

It looks like we never check whether the offset register is actually live (not clobbered) at optimization site. Add the check whether it is live at MBB entrance. The rest should have already been verified.

5. Fixes a bad codegen

For whatever reason we do transformation without checking if the value in register actually reaches the user. This is second identical fix for this pass.

   Co-authored-by: Anirudh Sundar <quic_sanirudh@quicinc.com>
   Co-authored-by: Sergei Larin <slarin@quicinc.com>

### CODE REVIEW

**Comment by quic-asaravan** [comment]
@SundeepKushwaha @iajbar @quic-sanirudh @JyotsnaVerma Can you pls review this?

File: llvm/test/CodeGen/Hexagon/addrmode-opt-nonreaching.mir [hunk 1]
```diff
@@ -0,0 +1,232 @@
+# It is not safe to make the transformation if the definition is killed by a call.
+# Use debug output for simplicity and test resilience.
+#
+# RUN: llc -march=hexagon -run-pass amode-opt %s -debug -o  %t_1.mir 2>&1 | FileCheck %s
+#CHECK: The LRExtReg reg {{.*}} is NOT live in to MBB
```
Comment (androm3da) [inline llvm/test/CodeGen/Hexagon/addrmode-opt-nonreaching.mir:5]: Does `-passes='print<live-vars>'` provide a more specific/stable matching opportunity versus the debug output?

Comment (androm3da) [reply llvm/test/CodeGen/Hexagon/addrmode-opt-nonreaching.mir:5]: Maintainers might be surprised if a change to the debug text wording causes a test failure here, for example.

**Comment by androm3da** [comment]
LGTM but is it possible @nathanchance to confirm it addresses the compile-time issue?

**Review by androm3da — APPROVED** [review]

**Comment by nathanchance** [comment]
> LGTM but is it possible @nathanchance to confirm it addresses the compile-time issue?

I’ll double check and get back to you by the end of the day.

**Comment by nathanchance** [comment]
This appears to resolve the compile time impact for me. Just running `make` to build that one `.o` file takes over 9 minutes before this change and after this change, it only takes 30 seconds. Thanks a lot for the fix!

**Review by iajbar — APPROVED** [review]

**Pull request merged**